### ISSUE TITLE
Custom fields on custom metrics

### ISSUE THREAD

**Issue by elFarto** [stagemonitor/stagemonitor#177]
I can't see any way of adding custom fields to a custom metric. Currently I've got 3 related values per customer that I log as separate metrics. Unfortunately this makes it almost impossible to visualise in Kibana (issues aggregating over the value field, it not showing me one of the tag fields, etc..).

It would be nice if I could just set my own fields on each metric, rather than attempting to fit into the Gauge/Counter/Histogram/Meter/Timer pigeon holes.

I hacked a partial solution by abusing the tags on the MetricName class to allow me to return Objects, rather than just Strings, that I could then change when the Gauge was read. This worked well, and Kibana was more than happy to let me use those values. However it is a complete hack and not something I'd want to use in production.

Is it possible to get the ability to augment a Metric (perhaps with another Interface) to allow an arbitrary set of fields to be added? I can work on an implementation if it's something you're willing to accept.

**Comment by felixbarny** [comment]
Could you elaborate a bit more about your usecase, what kind of metrics you want to track and why they can't br visualized in Kibana?

**Comment by elFarto** [comment]
One of the issues is something you reported to the elastic.co guys regarding Kibana not letting you aggregate non-indexed fields, there's not much you can do about that, although it looks like it's going to be fixed for Kibana 5.

The other issue is that the data I want to log is made of 3 parts, a maximum, a 'soft-limit' value, and the current value. All of these values are logged per-customer. Currently I'm just using 3 Counters per customer, with each Counter having a tag denoting which type of value it holds, and another tag with the customer. But when it comes to visualising it as say a line-chart, the Y-axis can only be an aggregate of the 'value' field, which contains all 3 types of values.

I can't add more Y-axis values as there's no mechanism to split by term for the axis. I can split the lines base on the type of value (which is probably only happening because I used the 'type' field) which almost gets me there, but I'm unable to then split the chart by customer as the customer field doesn't show up in the list of fields.

I've hardly used Kibana/Elasticsearch so I'm probably missing something, and this might all be caused by Kibana bugs not showing me all the fields that the data has. However, even if Kibana did show all the fields, I'd still be limited to 1 type of aggregation on the Counter values, rather than the separate aggregation I can use when they're separate fields. Also, splitting the X-axis by the type term is not ideal as It limits the naming to the contents of the field and can't be customised.

Oddly I've just gone back to the visualisation I did that used the fields I hacked in, and something has changed, and it didn't work until I refreshed the field list.

**Comment by felixbarny** [comment]
One problem is that i can't easily add a custom metric type, because they are provided by the metrics library.

Have you tried Grafana? I think setting up that line chart should be possible with it. If you could post some sample documents and a quick paint mockup of how the graph should look like I can help you with this.

**Comment by elFarto** [comment]
I realise that adding a complete new type of Metric would be difficult, however it should be possible to have an addition interface on an existing Metric that the ElasticsearchReporter class could check for when writing out values. Something like this:

public interface AdditionalMetrics {
	public void addMetrics(JsonGenerator jg);
}

This would allow any type of Metric to add any fields it liked to the output, although I'm not too sure passing in the JsonGenerator is a great idea, maybe returning a Map<String, Object> would be better.

Here's how the graph looks on my test metric. The test data isn't great, the available is stuck on the same line as the limit, so you can't see it easily.

![image](https://cloud.githubusercontent.com/assets/2325080/16726621/197c789a-4755-11e6-8332-5d7b316054ee.png)


I have used Grafana, but I can't use it for this project as it will be using an existing ELK stack.

**Comment by felixbarny** [comment]
What about sending these kinds of metrics directly to Elasticsearch? That way you have complete control over them. `org.stagemonitor.core.elasticsearch.ElasticsearchClient#index` is your friend here.

**Comment by elFarto** [comment]
Using the ElasticsearchClient directly is a possible solution, but certainly not ideal. I would have to come up with my own logic for archiving data in the indexes, and for actually reporting it, which is all a duplication of the logic that stagemonitor all ready has.

**Comment by felixbarny** [comment]
How would you register the new metric types? The metric registry is currently not laid out to store anything but the main Metric types. I could change this, but I want to stay conceptually compatible with the MetricsRegistry from the Metrics library.

When writing your own reporter, there would actually not be a lot of duplication if you reuse the existing classes. Extend `ScheduledMetrics2Reporter` and initialize the builder with a `new Metric2Registry()` dummy. Then add a field `ConcurrentMap<MetricName, MyCustomMetric` and override the `report()` method. Now look at the implementation of `org.stagemonitor.core.metrics.metrics2.ElasticsearchReporter.reportMetrics`. You just need a different implementation for the OutputStreamHandler. The rest is the same: 

```
httpClient.send("POST", corePlugin.getElasticsearchUrl() + "/_bulk", null,
					metricsOutputStreamHandler, new ElasticsearchClient.BulkErrorReportingResponseHandler());
```

You don't even need you own index. Just reuse the index stagemonitor creates:

```
String bulkAction = ElasticsearchClient.getBulkHeader("index", STAGEMONITOR_METRICS_INDEX_PREFIX + StringUtils.getLogstashStyleDate(), METRICS_TYPE);
```

That way you can reuse the existing index management tasks.

**Comment by elFarto** [comment]
Sorry, I wasn't being clear before, the AdditionalMetrics interface would be in addition to one of the existing Metric interfaces. So you'd create a class that implements both Counter and AdditionalMetrics. However, the more I've looked into this, the more it seems attempting to reuse the metrics index for custom data is just a bad idea.

I've implemented my own index now, using the ElasticsearchClient class to do most of the work, and that seems to be ok.
